pkg/controller/dnsendpoint/dnsendpoint_controller.go

```diff
@@ -158,61 +161,60 @@ func (r *ReconcileDNSEndpoint) Reconcile(request reconcile.Request) (reconcile.R
 		return reconcile.Result{}, nil
 	}
 
-	desiredNameServers := sets.NewString(instance.Spec.Endpoints[0].Targets...)
+	desiredNameServers := sets.NewString(instance.Status.NameServers...)
 	rootDomain, currentNameServers := r.nameServerScraper.GetEndpoint(domain)
-	if currentNameServers.Equal(desiredNameServers) {
-		dnsLog.Debug("NS record is up to date")
-		return reconcile.Result{}, r.observeCurrentGeneration(instance, dnsLog)
-	}
 
-	dnsLog.Info("creating NS record")
-	if err := r.nameServerQuery.Create(rootDomain, domain, desiredNameServers); err != nil {
-		dnsLog.WithError(err).Error("error creating NS record")
-		return reconcile.Result{}, err
-	}
-	r.nameServerScraper.AddEndpoint(request.NamespacedName, domain, desiredNameServers)
-
-	return reconcile.Result{}, r.observeCurrentGeneration(instance, dnsLog)
-}
-
-func (r *ReconcileDNSEndpoint) syncDeletedEndpoint(instance *hivev1.DNSEndpoint, dnsLog log.FieldLogger) error {
-	dnsLog.Info("deleting NS record")
-	domain := instance.Spec.Endpoints[0].DNSName
-	rootDomain, currentNameServers := r.nameServerScraper.GetEndpoint(domain)
-	if err := r.nameServerQuery.Delete(rootDomain, domain, currentNameServers); err != nil {
-		dnsLog.WithError(err).Error("error deleting NS record")
-		return err
-	}
-	r.nameServerScraper.RemoveEndpoint(domain)
-	controllerutils.DeleteFinalizer(instance, hivev1.FinalizerDNSEndpoint)
-	if err := r.Update(context.Background(), instance); err != nil {
-		dnsLog.WithError(err).Log(controllerutils.LogLevel(err), "error deleting finalizer")
-		return err
+	switch {
```

Comment:
I find this switch a little bit hard to follow, I think because we go from two clauses inclusing !isDeleted and then switch to deleting for the default, looking at it I guess that's because we either have a deletionTimestamp or we have no desiredNameServers. A comment clearing this up might help. 

Also I can't immediately see what the situation is where we'd have no desired name servers, a comment there might be helpful as well.

Reply:
The scenario where there would be no desired name servers is if the hosted zone was deleted or lost for some reason and maybe could not be recreated. The DNSZone controller would clear the name servers from the status of the DNSZone resource. The DNSEndpoint controller at that point should delete the NS records since they are no longer relevant.

Reply:
Would still like to see a comment here just clarifying what's going on.